### ISSUE TITLE
connect: vault CA quickly fills up the 'connect-intermediates' backend with certtificates

### ISSUE THREAD

**Issue by cristicalin** [hashicorp/consul#4475]
#### Overview of the Issue

We configured a staging environment with connect (3 servers + ~20 nodes) with vault CA backend.
Over time (~1 day) the intermediate vault backend fills up with certificates causing vault to slow down and consul itself to crash due to OOM conditions because we were running vault with the same consul as its backend.

After ~1day when investigating the crash we noticed that the `connect-intermediate` backend vault backend had about 500000 certificates issued and the consul raft snapshots had grown to +4GB due to the extra data generated in that short time interval.

From my research into the documentation I was not able to identify any tunnable parameters for the vault CA backend configuration to adjust the TTL of the vault issued certs, since I had to clean up the environment I don't have any details with regard to how the certificates had been issued and what their TTL was.

This also seems to be a bit of a chicken and egg issue which should be documented because the bootstrapping of such and environment and full outage becomes a lot more complicated if vault uses the same consul cluster.

#### Reproduction Steps

Server configurations:

```
{
  "connect": {
    "enabled": true,
    "ca_provider": "vault",
    "ca_config": {
      "address": "http://vault.service.consul:8200",
      "token": "{{ vault_token }}",
      "root_pki_path": "connect-root",
      "intermediate_pki_path": "connect-intermediate"
    }
  }
}
```

Clients configuration:

```
[staging compute] root@lsi1hyp1001:~# cat /etc/consul/memcache-client.json 
{
  "service": {
    "name": "memcache-client",
    "port": 11211,
    "connect": {
      "proxy": {
        "config": {
	  "upstreams": [{
	    "destination_name": "memcache",
	    "local_bind_port": 11211
	  }]
	}
      }
    }
  }
}
[staging compute] root@lsi1hyp1001:~# cat /etc/consul/rabbitmq-client.json 
{
  "service": {
    "name": "rabbitmq-client",
    "port": 5673,
    "connect": {
      "proxy": {
        "config": {
	  "upstreams": [{
	    "destination_name": "rabbitmq",
	    "local_bind_port": 5673
	  }]
	}
      }
    }
  }
}
```

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
[staging compute] root@lsi1hyp1001:~# docker exec consul-agent consul info
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 4
	services = 6
build:
	prerelease = 
	revision = 28141971
	version = 1.2.0
consul:
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 32
	goroutines = 2450
	max_procs = 32
	os = linux
	version = go1.10.1
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 567
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 3663
	members = 44
	query_queue = 0
	query_time = 21
```

</details>

<details>
  <summary>Server info</summary>

```
[staging consul-server] root@lsm1kvs0001:~# docker exec consul-agent consul info
agent:
        check_monitors = 0
        check_ttls = 1
        checks = 3
        services = 3
build:
        prerelease = 
        revision = e716d1b5
        version = 1.2.2
consul:
        bootstrap = false
        known_datacenters = 1
        leader = false
        leader_addr = 172.18.0.79:8300
        server = true
raft:
        applied_index = 2510547
        commit_index = 2510547
        fsm_pending = 0
        last_contact = 79.968695ms
        last_log_index = 2510547
        last_log_term = 79
        last_snapshot_index = 2505154
        last_snapshot_term = 78
        latest_configuration = [{Suffrage:Voter ID:3cb4a8ed-5e58-2a2f-9bbc-b19d3267d8e6 Address:172.18.0.79:8300} {Suffrage:Voter ID:20a916f9-fb28-54dc-b05a-7c29fc50f04b Address:172.18.0.78:8300} {Suffrage:Voter ID:82b09ad5-8381-79cc-8c15-08c45a8215e3 Address:172.18.0.77:8300}]
        latest_configuration_index = 1
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 79
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 307
        max_procs = 2
        os = linux
        version = go1.10.1
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 567
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 3663
        members = 44
        query_queue = 0
        query_time = 21
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 10
        members = 3
        query_queue = 0
        query_time = 1
```

</details>

Server details above come after we recovered and disabled the vault CA so might not be relevant.

### Operating system and Environment details

OS: RHEL 7.5
Consul versions: 1.2.(0,1,2)
Running both server and agents in docker

**Comment by banks** [comment]
Hey @cristicalin 

Thanks for the report. This is certainly not intended behaviour! We'll work on a reproduction case and get this fixed.

> This also seems to be a bit of a chicken and egg issue which should be documented because the bootstrapping of such and environment and full outage becomes a lot more complicated if vault uses the same consul cluster.

Indeed bootstrapping Vault as a CA when it's using the same Consul cluster as it's backend has some issues. I think that's orthogonal to the bug here though - we should not be generating anywhere near that number of certificates. our default TTL for certificates is 3 days and we should be cleaning them up.

We will indeed have to find ways to improve the bootstrapping problem when it comes to using Connect to connect to Vault in the first place but that's a separate issue 😄.

> I was not able to identify any tunnable parameters for the vault CA backend configuration to adjust the TTL of the vault issued certs

We actually made the certificate TTL configurable in the release on Monday: https://www.consul.io/docs/agent/options.html#ca_leaf_cert_ttl but I don't think that is the issue here as you should not have more than 1 certificate per service instance with the default settings. (Agent restarts can cause new certificates to be issued so maybe a couple more than that but nothing like what you are seeing).

Another near-future enhancement will involve Consul servers tracking more closely meta-data about each certificate to enable revocation and part of that metadata will include whether the service is actually still using it which means we'll be able to clean up old certs more aggressively too rather than let them just die after they expire.

**Comment by banks** [comment]
Hi @cristicalin,

Would you be able to supply the log files from at least one server and one of the client agents please? Preferably with `log_level = DEBUG` but if you didn't have that enabled and have disabled vault now then please consider adding that to config for any future repro attempts!

It seems like somewhere, something is in a busy loop and not the expected cycle for certificate updates - 500k certificates in a day means it's regenerating them on average about 6 per second! You have 20 client nodes each with 2 services on (assuming both the services you showed are on each node) so you should have about 40 certificates and they should only change every few days so after a few days I'd expect to see 80 or maybe 120 before the oldest ones expire.

Logs may well help identify where the busy loop is, particularly interested to see if there is some error condition that is causing repeated Sign requests and somehow bypassing normal waiting.

We ran a cluster with the same configuration as yours overnight and identified one or two minor issues but nothing like what you saw yet.

Also I note you client is `1.2.0`. Could you please try upgrading them all to 1.2.2 and enabling Vault again? We've fixed a few issues with error conditions in edge cases in proxy and agent code so I'd like to rule those out too.

Please note that in addition to restarting the client agents, **you will also need restart the managed proxies manually** since they intentionally live on through agent restarts/crashes.

The simplest way to do this currently is to restart the agent first and then find the proxies and kill them manually something like `ps aux | grep 'consul connect proxy` then `kill` the processes with TERM. The same `ps` command should show two _new_ ones with different PIDs come back since the agent should notice and restart them pretty instantly.

You should see 2 on each client assuming both your services run on each client node (in general one proxy for every service instance registered on the node). You could also temporarily remove the service definitions and HUP so the proxies get unregistered and stopped by the agent and then add them back, but please check with `ps` that only the desired number of proxy processes are running after this (should be true but don't want potential proxy management bugs to confuse this diagnosis!)

We plan to make this process simpler in the future!

If you are happy to post logs in a gist that would be great, if there is anything you consider sensitive in them and you are more comfortable sharing them privately then let me know and I will find a suitable place for that.

Thanks!

**Comment by pearkes** [comment]
We no longer store certificates in Vault so this should be fixed in the 1.4.0-rc1 https://releases.hashicorp.com/consul/1.4.0-rc1. Please let us know if you see the same behavior.

If you want to open a separate issue about the OOM problems (which do sound like a potential issue) that'd be great.
